confidential-data-hub/storage/src/volume_type/rbd/mod.rs

```diff
@@ -0,0 +1,156 @@
+use log::{debug, error};
+use rand::{distributions::Alphanumeric, Rng};
+use serde::{Deserialize, Serialize};
+use std::collections::HashMap;
+use strum::{Display, EnumString};
+use tokio::{
+    fs,
+    io::{AsyncReadExt, AsyncWriteExt},
+    process::Command,
+};
+
+/// LUKS encrypt storage binary
+const LUKS_ENCRYPT_STORAGE_BIN: &str = "/usr/local/bin/luks-encrypt-storage";
+
+#[derive(EnumString, Serialize, Deserialize, Display, Debug, PartialEq, Eq)]
+pub enum RbdEncryptType {
+    #[strum(serialize = "luks")]
+    LUKS,
+}
+
+#[derive(Serialize, Deserialize, PartialEq, Debug)]
+struct RbdParameters {
+    /// The device number, formatted as "MAJ:MIN".
+    #[serde(rename = "deviceId")]
+    pub device_id: String,
+
+    /// The encryption type. Currently, only LUKS is supported.
+    #[serde(rename = "encryptType")]
+    pub encrypt_type: RbdEncryptType,
+
+    /// Encryption key. If not set, generate a random 4096-byte key
+    #[serde(rename = "encryptKey")]
+    pub encrypt_key: Option<String>,
+
+    /// Indicates whether to enable dm-integrity.
+    #[serde(rename = "dataIntegrity")]
+    pub data_integrity: String,
+}
+
+pub(crate) struct Rbd;
+
+async fn random_encrypt_key() -> anyhow::Result<String> {
+    let mut buffer = vec![0u8; 4096];
+    rand::thread_rng().fill(&mut buffer[..]);
+    Ok(base64::engine::general_purpose::STANDARD.encode(&buffer))
+}
+
+async fn get_plaintext_secret(secret: &str) -> anyhow::Result<String> {
+    if secret.starts_with("sealed.") {
+        debug!("detected sealed secret");
+        let unsealed = secret::unseal_secret(secret.as_bytes()).await?;
+
+        String::from_utf8(unsealed).context("convert to String failed")
+    } else {
+        Ok(secret.into())
+    }
+}
+
+async fn create_storage_key_file(rbd_parameter: &RbdParameters) -> Result<String> {
+    let random_string: String = rand::thread_rng()
+        .sample_iter(&Alphanumeric)
+        .take(5)
+        .map(char::from)
+        .collect();
+    let storage_key_path = format!("/tmp/encrypted_disk_{}", random_string);
+    let mut storage_key_file = fs::File::create(storage_key_path.clone()).await?;
+
+    let plain_key = match &rbd_parameter.encrypt_key {
+        Some(encrypt_key) => get_plaintext_secret(encrypt_key).await?,
+        None => random_encrypt_key().await?,
+    };
+
+    storage_key_file.write_all(plain_key.as_bytes()).await?;
+    storage_key_file.flush().await?;
+    Ok(storage_key_path)
+}
+
+impl Rbd {
+    async fn real_mount(
+        &self,
+        options: &HashMap<String, String>,
+        _flags: &[String],
+        mount_point: &str,
+    ) -> Result<()> {
+        // construct RbdParameters
+        let parameters = serde_json::to_string(options)?;
+        let rbd_parameter: RbdParameters = serde_json::from_str(&parameters)?;
+
+        if rbd_parameter.encrypt_type == RbdEncryptType::LUKS {
+            let storage_key_path = create_storage_key_file(&rbd_parameter).await?;
```

Comment:
should we delete this file at some point?

Reply:
It would be removed in the binary `luks-encrypt-storage`